chain/network/src/routing/edge_cache/mod.rs

```diff
@@ -0,0 +1,347 @@
+                    false
+                }
+            }
+            im::hashmap::Entry::Vacant(_) => {
+                assert!(false);
+                false
+            }
+        };
+        if is_newly_inactive {
+            self.decrement_degrees_for_key(key);
+        }
+    }
+
+    /// Stores the key-value pair (peer_id, edges) in the EdgeCache's `active_trees` map, overwriting
+    /// any previous entry for the same peer. Updates `active_edges` accordingly.
+    pub fn update_tree(&mut self, peer_id: &PeerId, tree: &Vec<Edge>) {
+        // Insert the new edges before removing any old ones.
+        // Nodes are pruned from the `p2id` mapping as soon as all edges incident with them are
+        // removed. If we removed the edges in the old tree first, we might unlabel and relabel a
+        // node unnecessarily. Inserting the new edges first minimizes churn on `p2id`.
+        for edge in tree {
+            self.insert_active_edge(edge);
+        }
+
+        let edge_keys: Vec<EdgeKey> = tree.iter().map(|edge| edge.key().into()).collect();
+
+        if let Some(old_edge_keys) = self.active_trees.insert(peer_id.clone(), edge_keys) {
+            // If a previous tree was present, process removal of its edges
+            for key in &old_edge_keys {
+                self.remove_active_edge(key);
+            }
+        }
+    }
+
+    /// Removes the tree stored for the given peer, if there is one.
+    pub fn remove_tree(&mut self, peer_id: &PeerId) {
+        if let Some(edges) = self.active_trees.remove(peer_id) {
+            for e in &edges {
+                self.remove_active_edge(e);
+            }
+        }
+    }
+
+    /// Upper bound on mapped u32 ids; not inclusive
+    pub fn max_id(&self) -> usize {
+        self.degree.len()
+    }
+
+    /// Iterator over the (PeerId, u32) mapping
+    pub fn iter_peers(&self) -> Iter<'_, PeerId, u32> {
+        self.p2id.iter()
+    }
+
+    /// Number of known edges in the network
+    pub fn known_edges_ct(&self) -> usize {
+        self.verified_nonces.len()
+    }
+
+    /// Prunes entries with nonces older than `prune_nonces_older_than`
+    /// from `verified_nonces`
+    pub fn prune_old_edges(&mut self, prune_nonces_older_than: u64) {
+        // Drop any entries with old nonces from the verified_nonces cache
+        self.verified_nonces.retain(|_, nonce| nonce >= &prune_nonces_older_than);
+    }
+
+    /// Accepts a mapping over the set of reachable PeerIds in the network
+    /// to the shortest path lengths to those peers.
+    ///
+    /// Constructs a tree from among the `active_edges` which has the same
+    /// reachability and the same distances or better.
+    ///
+    /// Returns None if the input is inconsistent with the state of the cache
+    /// (reachability or distances are not consistent with the `active_edges`).
+    pub fn construct_spanning_tree(&self, distance: &HashMap<PeerId, u32>) -> Option<Vec<Edge>> {
```

Comment:
I'm a bit confused about this method. It doesn't seem to verify that the result is indeed a spanning tree. Do we trust the arguments here?

Reply:
The only implicit assumption here is that the local node is present in the distance mapping and has distance 0. I will add an explicit check for that condition.

Let `L` be the local node. If all of these conditions are satisfied:
- `distance`'s values are non-negative integers
- `distance[L] == 0`
- For every node N != L in distance's key-set, we include an edge `(N, M)` such that `distance[N] > distance[M]`

that is sufficient to conclude that we have constructed a spanning tree achieving all of the distances specified. It is sort of an inductive argument; if each non-local node has an edge which brings us closer to the local node, we can take those edges one by one to construct a path from any node to the local node. It is easy to see that such paths achieve the claimed distances. We have a connected graph on `distance.len()` nodes using `distance.len() - 1` edges, so it's a tree.

Reply:
Thanks! I see it now. Makes sense with the decreasing distance to root. I initially missed the fact that cycles are impossible thanks to the fact that for an edge to be added it needs to go to a node with **strictly** smaller distance. I finally convinced myself with the following argument. There are no cycles so we have a spanning forest plus some number of nodes without any edges. The only way to get the #edges as high as #nodes + 1 is if the forest only contains one tree and no nodes without edges.